launch/launch/launch_service.py

```diff
@@ -290,7 +290,10 @@ def _on_exception(loop, context):
                     # the queue
                     is_idle = self._is_idle()  # self._entity_future_pairs is pruned here
                     if not self.__shutting_down and shutdown_when_idle and is_idle:
-                        ret = await self._shutdown(reason='idle', due_to_sigint=False)
+                        coroutine = self._shutdown(reason='idle', due_to_sigint=False)
+                        ret = None
+                        if coroutine is not None:
+                            ret = await coroutine()
```

Comment:
shouldn't it be
```suggestion
                            ret = await coroutine
```
?

Reply:
From my search online, I think you await a _coroutine_ with parentheses (it may take arguments) and you await a _task_ without the parentheses.

Running the following code seems to confirm part of that:

```
import asyncio

async def bar():
    print('bar')

async def foo():
    await bar

asyncio.run(foo())
```

Results in an error:

```
TypeError: object function can't be used in 'await' expression
```

Reply:
`bar` and `foo` are a coroutine function, that return a coroutine object.
You need to await coroutines, and call coroutine functions before awaiting on the coroutine object.

From python [docs](https://docs.python.org/3/library/asyncio-task.html#awaitables):

> Important In this documentation the term “coroutine” can be used for two closely related concepts:
> a coroutine function: an [async def](https://docs.python.org/3/reference/compound_stmts.html#async-def) function;
> 
> a coroutine object: an object returned by calling a coroutine function.

---

Before the last [commit](https://github.com/ros2/launch/pull/603/commits/7a7d5569b578cd6c7277c17c2e59ce09cbb80b2b), `_shutdown()` was returning a coroutine object, so I think `ret = await coroutine` was the correct thing (probably, calling the coroutine object was generated the warning).
After that commit we're returning a coroutine function, so `ret = await coroutine()` is now fine

Reply:
Ah, thanks for the clarification! That explains the odd behavior I was seeing.  Sorry, I'm not that familiar with asyncio.

I've reverted [7a7d556](https://github.com/ros2/launch/pull/603/commits/7a7d5569b578cd6c7277c17c2e59ce09cbb80b2b) and avoided calling the coroutine ([8a47880](https://github.com/ros2/launch/pull/603/commits/8a47880eb0ce51ff1192655d676b0b9acbdc5184)),